clients/pkg/promtail/wal/watcher.go

```diff
@@ -191,53 +195,76 @@ func (w *Watcher) watch(segmentNum int) error {
 			// return after reading the whole segment for creating a new LiveReader from the newly created segment
 			return nil
 
-		case <-readTicker.C:
-			err = w.readSegment(reader, segmentNum)
-			if debug {
-				level.Warn(w.logger).Log("msg", "Error reading segment inside readTicker", "segment", segmentNum, "read", reader.Offset(), "err", err)
-			}
+		// the cases below will unlock the select block, and execute the block below
+		// https://github.com/golang/go/issues/23196#issuecomment-353169837
+		case <-w.readNotify:
+		case <-readTimer.C:
+		}
 
-			// io.EOF error are non-fatal since we are tailing the wal
-			if errors.Cause(err) != io.EOF {
-				return err
-			}
+		// read from open segment routine
+		ok, err := w.readSegment(reader, segmentNum)
+		if debug {
+			level.Warn(w.logger).Log("msg", "Error reading segment inside readTicker", "segment", segmentNum, "read", reader.Offset(), "err", err)
 		}
+
+		// io.EOF error are non-fatal since we are tailing the wal
+		if errors.Cause(err) != io.EOF {
+			return err
+		}
+
+		if ok {
+			// TODO: we'll be resetting the readTimer for each notification we get. Should we do this just if the readSegment
```

Comment:
only open doubt, should we differentiate this block when it was triggered by a notification from the writer, or the timer? Maybe not, since the timer is a backup that `attempts to read if a notification hasn't arrived for X time`

Reply:
The main question here is because if we don't differentiate, the timer would be "refreshed" every time we receive a write notification, which could happen often and could lead to a lot of unnecessary calls to the timer funcs that perform the refresh. Maybe worth benchmark + profiling?

wdyt @cstyan ?

Reply:
imo we should, if only just to include debug logging and/or a metric about timer reads

similarly it may also be useful to include a metric about dropped notifications in `NotifyWrite` (due to the channel already being full with it's 1 element)

Reply:
Good idea! Added both log and a metrics on timer-triggered segment reads, and a metrics on dropped notifications